### PR TITLE
WIP: mavlink GPS_INPUT support

### PR DESCRIPTION
**Pull request by dagar**

### CODE REVIEW

**Comment by Antiheavy** [comment]
Related comment in the dual GPS PR:
https://github.com/PX4/Firmware/pull/9765#issuecomment-404689066

**Comment by Antiheavy** [comment]
One issue that came up is how to handle GPS modules that do not send the expected altitude datum format for PX4/EKF2.  For example, we might use a GPS receiver that only outputs wgs84 altitude, this is a problem for PX4/EKF2 which is expecting whatever uBlox "amsl" datum is (I think egm96?).

- One idea is to send the MAVlink message to ignore the GPS altitude from the secondary MAVlink GPS source (param 1): https://mavlink.io/en/messages/common.html#GPS_INPUT_IGNORE_FLAGS

- Another idea is that we send a artificially very large EPV value from the secondary MAVlink GPS source so that the driver and/or estimator will effectively ignore the GPS altitude, but use the other information from the secondary GPS.

Both of the above ideas would only be useful when switching to the secondary GPS, the primary GPS would need to have proper altitude and everything else for system initialization, pre-flight checks, pre-arm checks, etc.

Another approach could be to add a parameter to select which uBlox altitude value to use.  Default to the current (egm96?), but allow an advanced user to select the wgs84 altitude so that it could match the wgs84 altitude on a secondary GPS module.

@jonrwat @priseborough @dagar

**Comment by LorenzMeier** [comment]
Could you please rebase? Thanks!

**Comment by jonrwat** [comment]
Several updates are recommended:

1. Scaling of values is incorrect (e.g. vel-ned).  It looks like the majority of the GPS_INPUT handler was copied from the HIL_GPS handler, which has a different scaling of data.  Scaling of parameters is specified at: [https://github.com/mavlink/c_library_v2/blob/master/common/mavlink_msg_gps_input.h#L7](https://github.com/mavlink/c_library_v2/blob/master/common/mavlink_msg_gps_input.h#L7)

2. All data has not been copied to the "vehicle_gps_position" structure and published (e.g. hdop, vdop).  This is apparent when looking at GPS status in QGC (both GUI display and MAVLink Inspector), as fields are missing (i.e. locked as zero).

3. The software doesn't use the GPS_INPUT message 'ignore_flags' field.  This causes the system (and EKF) to potentially use data which is invalid.  This likely isn't as great of a concern on platforms where all source data is supplied in a single message (e.g. u-blox UBX_NAV_PVT) but has greater concern on platforms where source data is fragmented across multiple messages (e.g. Swift Piksi).  I'm not sure what the final solution is; but it seems harmful to feed known invalid data into the EKF.

**Comment by TSC21** [comment]
Jenkins test this please.

**Comment by TSC21** [comment]
@dagar maybe we can rebase this on master?

@jonrwat we are probably bringing this as is now. We thank you if you can issue a PR with the changes you propose after it. Thanks!

**Comment by dagar** [comment]
Rebased on master. We could probably do a quick pass and get this in safely.

**Comment by Antiheavy** [comment]
Still a valid issue

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 1]
```diff
@@ -2174,6 +2178,44 @@ MavlinkReceiver::handle_message_hil_sensor(mavlink_message_t *msg)
...
+
+	gps.timestamp_time_relative = 0;
+	gps.time_utc_usec = gps_input.time_usec;
+
+	gps.timestamp = timestamp;
+	gps.lat = gps_input.lat;
+	gps.lon = gps_input.lon;
+	gps.alt = gps_input.alt;
+	gps.eph = (float)gps_input.horiz_accuracy * 1e-2f; // from cm to m
+	gps.epv = (float)gps_input.vert_accuracy * 1e-2f; // from cm to m
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_receiver.cpp]: The spec says this is all meters, not cm.

https://mavlink.io/en/messages/common.html#GPS_INPUT

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 2]
```diff
@@ -2174,6 +2178,44 @@ MavlinkReceiver::handle_message_hil_sensor(mavlink_message_t *msg)
...
+
+	gps.s_variance_m_s = 1.0f;
+
+	//gps.vel_m_s = (float)gps_input.vel * 1e-2f; // from cm/s to m/s
+	gps.vel_n_m_s = gps_input.vn * 1e-2f; // from cm to m
+	gps.vel_e_m_s = gps_input.ve * 1e-2f; // from cm to m
+	gps.vel_d_m_s = gps_input.vd * 1e-2f; // from cm to m
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_receiver.cpp]: Same here.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 3]
```diff
@@ -2174,6 +2178,44 @@ MavlinkReceiver::handle_message_hil_sensor(mavlink_message_t *msg)
...
+	gps.vel_ned_valid = true;
+	//gps.cog_rad = ((gps_input.cog == 65535) ? NAN : wrap_2pi(math::radians(gps_input.cog * 1e-2f)));
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_receiver.cpp]: That's not a thing right?

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 4]
```diff
@@ -2174,6 +2178,44 @@ MavlinkReceiver::handle_message_hil_sensor(mavlink_message_t *msg)
...
+
+	gps.fix_type = gps_input.fix_type;
+	gps.satellites_used = gps_input.satellites_visible;  //TODO: rename mavlink_hil_gps_t sats visible to used?
```
Comment (julianoes) [inline src/modules/mavlink/mavlink_receiver.cpp]: Too late now, I think.

**Comment by LorenzMeier** [comment]
Stale, closing.

**Pull request closed without merging**